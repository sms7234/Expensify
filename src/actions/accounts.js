import database from '../firebase/firebase';

export const setAccounts = (accounts) => ({
  type: "SET_ACCOUNTS",
  accounts
});

export const startSetAccounts = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/accounts`)
      .once('value')
      .then((snapshot) => {
        const accounts = [];
        snapshot.forEach((child) => {
          accounts.push({
            id: child.key,
            ...child.val()
          });
        });
        dispatch(setAccounts(accounts));
      });
  };
};

const addAccount = (account) =>({
  type: "ADD_ACCOUNT",
  account
});

export const startAddAccount = (accountData={}) => {
  return (dispatch, getState) => {
    const {
      account='',
      description=''
    } = accountData;
    const accountItem={account,description};
    const uid=getState().auth.uid;
    return database.ref(`users/${uid}/accounts`).push(accountItem).then((ref)=>{
      dispatch(addAccount({
        id: ref.key,
        ...accountItem
      }));
    });
  };
};

export const editAccount = (id, updates) => ({
  type: 'EDIT_ACCOUNT',
  id,
  updates
});

export const startEditAccount = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/accounts/${id}`).update(updates).then(()=>{
      dispatch(editAccount(id,updates));
    });
  };
};

export const removeAccount = (id) => ({
  type: 'REMOVE_ACCOUNT',
  id
});

export const startRemoveAccount = ({ id } = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/accounts/${id}`).remove().then(()=>{
      dispatch(removeAccount(id));
    });
  };
};
