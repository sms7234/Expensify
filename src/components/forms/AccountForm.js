import React from 'react';

export default class AccountForm extends React.Component{
  constructor(props){
    super(props);
    this.state={
      account: props.accounts ? props.accounts.account: '',
      description: props.accounts ? props.accounts.description:'',
      error:''
    };
  }
  onAccountChange = (e) => {
    const account = e.target.value;
    if(account.length > 0){
      this.setState(()=> ({account}));
    };
  };
  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(()=>({description}))
  };
  onSubmit = (e) => {
    e.preventDefault();
    if(!this.state.account){
      this.setState(()=> ({error: 'Please fill in account'}));
    } else {
      this.setState(()=>({error:''}));
      this.props.onSubmit({
        account: this.state.account,
        description: this.state.description
      })
    }
  };
  render() {
    return (
      <form className="form" onSubmit={this.onSubmit}>
        {this.state.error!=='' && <p className="form__error">{this.state.error}</p>}
        <input
          type="text"
          placeholder="account name"
          className="text-input"
          value={this.state.account}
          onChange={this.onAccountChange}
          autoFocus
        />
        <textarea
          placeholder="Add description for your account"
          className="textarea"
          value={this.state.description}
          onChange={this.onDescriptionChange}
        />
        <button className="button">
          Save
        </button>
      </form>
    )
  }
}
