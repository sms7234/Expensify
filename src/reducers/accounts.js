export default (state=[], action) => {
  switch(action.type){
    case 'SET_ACCOUNTS':
      return action.accounts;
    case 'ADD_ACCOUNT':
      return state.concat(action.account);
    case 'EDIT_ACCOUNT':
      return state.map((item) => {
        if (item.id === action.id){
          return {
            ...item,
            ...action.updates
          }
        } else {return item}
      });
      case 'REMOVE_ACCOUNT':
        return state.filter(({id}) => id!==action.id);
    default:
      return state;
  }
}
