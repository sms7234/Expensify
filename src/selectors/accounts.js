export default (accounts, {account}) => {
  return accounts.filter((item) => {
    const textMatch = item.account.toLowerCase().includes(account.toLowerCase());
    return textMatch;
  }).sort((a,b)=> {
    if(a.account < b.account){
      return -1;
    } else{
      return 1;
    }
  });
};
