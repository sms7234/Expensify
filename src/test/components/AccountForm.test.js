import React from 'react';
import {shallow} from 'enzyme';
import AccountForm from '../../components/AccountForm';
import accounts from '../fixtures/accounts';

test('should render Account Form with no data', () => {
  const wrapper = shallow(<AccountForm />);
  expect(wrapper).toMatchSnapshot();
});

test('should render AccountForm with data', () => {
  const wrapper = shallow(<AccountForm account={accounts[0]}/>);
  expect(wrapper).toMatchSnapshot();
});

test('should render error for invalid submission', () => {
  const wrapper = shallow(<AccountForm />);
  expect(wrapper).toMatchSnapshot();
  wrapper.find('form').simulate('submit', {
    preventDefault: () => { }
  });
  expect(wrapper.state('error').length).toBeGreaterThan(0);
  expect(wrapper).toMatchSnapshot();
});

test('should set account value on change', () => {
  const value = "new account";
  const wrapper= shallow(<AccountForm />);
  wrapper.find('input').simulate('change', {
    target: {value}
  });
  expect(wrapper.state('account')).toBe(value);
});

test('should set account value on change', () => {
  const value = "new description";
  const wrapper= shallow(<AccountForm />);
  wrapper.find('textarea').simulate('change', {
    target: {value}
  });
  expect(wrapper.state('description')).toBe(value);
});
