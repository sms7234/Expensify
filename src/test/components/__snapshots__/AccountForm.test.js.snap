// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Account Form with no data 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <input
    autoFocus={true}
    className="text-input"
    onChange={[Function]}
    placeholder="account name"
    type="text"
    value=""
  />
  <textarea
    className="textarea"
    onChange={[Function]}
    placeholder="Add description for your account"
    value=""
  />
  <button
    className="button"
  >
    Save
  </button>
</form>
`;

exports[`should render AccountForm with data 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <input
    autoFocus={true}
    className="text-input"
    onChange={[Function]}
    placeholder="account name"
    type="text"
    value=""
  />
  <textarea
    className="textarea"
    onChange={[Function]}
    placeholder="Add description for your account"
    value=""
  />
  <button
    className="button"
  >
    Save
  </button>
</form>
`;

exports[`should render error for invalid submission 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <input
    autoFocus={true}
    className="text-input"
    onChange={[Function]}
    placeholder="account name"
    type="text"
    value=""
  />
  <textarea
    className="textarea"
    onChange={[Function]}
    placeholder="Add description for your account"
    value=""
  />
  <button
    className="button"
  >
    Save
  </button>
</form>
`;

exports[`should render error for invalid submission 2`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <p
    className="form__error"
  >
    Please fill in account
  </p>
  <input
    autoFocus={true}
    className="text-input"
    onChange={[Function]}
    placeholder="account name"
    type="text"
    value=""
  />
  <textarea
    className="textarea"
    onChange={[Function]}
    placeholder="Add description for your account"
    value=""
  />
  <button
    className="button"
  >
    Save
  </button>
</form>
`;
